src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java

```diff
@@ -1360,4 +1363,39 @@ public boolean shouldTick() {
     public void setInvulnerable(boolean value) {
         this.invulnerable = value;
     }
+
+    @Inject(method = "setFire", at = @At(value = "HEAD"), cancellable = true)
+    public void onFire(int seconds, CallbackInfo ci) {
+        if (((IMixinWorld) world).isFake()) {
+            return;
+        }
+        if (!this.isBurning()) {
+            try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+                if (this.originalLava != null) {
+                    frame.pushCause(this.originalLava);
+                }
+                if (this.originalInFire != null) {
+                    frame.pushCause(this.originalInFire);
+                }
+                if (this.originalLightning != null) {
+                    frame.pushCause(this.originalLightning);
+                }
+
+                frame.pushCause(this.getLocation().getExtent());
+                IgniteEntityEvent event = SpongeEventFactory.
+                        createIgniteEntityEvent(frame.getCurrentCause(), this.fire, this.fire, this);
+
+                if (SpongeImpl.postEvent(event)) {
+                    ci.cancel();
+                }
+            }
+        }
+    }
+
+    private DamageSource originalLightning;
+
+    @Inject(method = "onStruckByLightning", at = @At(value = "HEAD"))
```

Comment:
Where is this coming from?

Reply:
This was created to add the DamageSource when the entity is set on fire by a lightning. Previously there was a field called `originalLava` (set inside `setOnFireFromLava()`) for lava and a field called `originalInFire` (set inside `dealFireDamage`) for fire, but nothing for lightning.

Reply:
If it's coming from that, you need to be verifying where the modified lightning is being set. If it isn't being set, then this isn't doing anything special.

Reply:
Here's the thing, what I tried asking indirectly is that your'e not setting the `originalLightning` back to `DamageSource.LIGHTNING_BOLT`, nor are you making it a new `DamageSource`, so this injection is 100% useless without that logic existing to *reset* the damage sources.